### PR TITLE
fix: pass username with dagger modules from remote http ref

### PR DESCRIPTION
**Pull request by skycaptain**
In #8989, an issue emerged with GitLab CI's `CI_JOB_TOKEN`s: Right now, when accessing remote modules via HTTP authentication using `dagger call -m gitlab.com/...`, the username is hardcoded to `x-access-token` (except for bitbucket.com). However, as specified in [GitLab’s documentation](https://docs.gitlab.com/ci/jobs/ci_job_token/#use-a-job-token), username `gitlab-ci-token` is required when using basic authentication with `CI_JOB_TOKEN`. In the discussion, we recognized that not only receiving the secret but also the username from the git credential helper would solve the issue best.

Several potential solutions came to my mind:

- Add another option `httpAuthUsername` to `RemoteGitRepository` and pass the username alongside `httpAuthToken`. It would also change the interface of [`GitRepository`](https://docs.dagger.io/api/reference/#definition-GitRepository), and while this isn't related to the original problem, it makes it easier to set the username when calling a git repo from within a module (currently, the [`httpAuthHeader`](https://docs.dagger.io/api/reference/#query-git) arg must be composed manually). This was my first approach until I realized I could solve the issue without extensive code refactoring. However, I was considering adding two helper functions to [`GitRepository`](https://docs.dagger.io/api/reference/#definition-GitRepository): `.WithHttpAuthHeaderFromBasicAuth(username, password)` and `.WithHttpAuthHeaderFromBearerAuth(token)` to simplify [`httpAuthHeader`](https://docs.dagger.io/api/reference/#query-git) formatting.
- Use `httpAuthHeader` instead of `httpAuthToken` to leverage existing code with minimal refactoring. While this solution works for the PR, it reveals overlapping logic between `core.RemoteGitRepository.setup` and `core.schema.gitSchema.git`. Further refactoring may be beneficial to address this duplication.

Currently, this PR doesn't address another desirable CLI feature: being able to pass credentials within the URL. At present, username and password parameters in URLs are ignored to best of my knowledge. Adding this capability would allow using the format `dagger call -m https://gitlab-ci-token:${CI_JOB_TOKEN}@gitlab.com/...` and also passing repositories as directories from the command line in general. This would eliminate the need to set up a git credential helper, which can be cumbersome in some CI environments. However, during implementation, I discovered that additional refactoring is needed to make this work. Therefore, this PR only addresses the GitLab CI issue from #8989.

Finally, do you have any suggestions for how to write test for these changes?

### CODE REVIEW

**Comment by grouville** [comment]
Thanks for this PR 🙏 Given that all the tests pass, it seems to be a correct implementation. I'll dig a bit more tomorrow on the implications (haven't thoroughly checked yet). I'll add a corresponding test too in the meantime.

Privileges on the test repo are mandatory for this test ahah 😇 

Thanks again !

PS: regarding having the PAT inside the ref, I worry that we couldn't ensure 0 leak of information in cache the token gets cached

**Comment by skycaptain** [comment]
> [...] I'll dig a bit more tomorrow on the implications (haven't thoroughly checked yet). [...]

Sure! I absolutely understand that a change like this needs to be thought through and tested thoroughly to avoid breaking existing functionality or leaking sensitive data.

> I'll add a corresponding test too in the meantime.

Thanks!

**Comment by jedevc** [comment]
LGTM, we need a test though.

@grouville why do we need to make changes to the test repo here? This should just be possible to test just using `ExperimentalServiceHost`.

There's no tests that cover the behavior of the auto-loading credentials for just plain git clones (only for modules). https://github.com/dagger/dagger/blob/84dc5b3ba37238d957db1025c70dca797d35a055/core/integration/module_test.go#L5774-L5800

We should adapt the above to `git_test.go` and add tests for this through there.

File: core/schema/git.go [hunk 1]
```diff
@@ -349,8 +353,8 @@ func (s *gitSchema) git(ctx context.Context, parent dagql.Instance[*core.Query],
 					Value: dagql.NewString(remote.String()),
 				},
 				{
-					Name:  "httpAuthToken",
```
[the remark this answers is not in the document]
Comment (grouville) [reply core/schema/git.go:352]: I'm totally ok with adding the `httpAuthUsername` field to Git. As it changes the API, do you think we shall include Solomon for the API design decision ?

**Comment by grouville** [comment]
@jedevc I'm not 100% sure to understand all the implications and recommendations here 🙏 👼 

Just to make sure, this is what I understood:
> @grouville why do we need to make changes to the test repo here? This should just be possible to test just using ExperimentalServiceHost.
1. As this PR solves a very specific pain point: JOB_CI_TOKEN on GitLab do not work without the user setting the proper username -- I was asking whether we should do an integration test using a real token of that kind (and adding a test to `git_test.go` indeed)

But, what I understand is that: we should spin up a local git server, and, using `ExperimentalServiceHost`, we check locally if the username and token is correctly passed to the git server ? So, more like a unit test instead of an integration test on this very specific edge case. Just wanna double check 🙏 

2. I am also a bit confused by this sentence: `There's no tests that cover the behavior of the auto-loading credentials for just plain git clones (only for modules).`

What do you mean by `plain git clones` ? I double verified the fact that calls to `dag.Git()` inside a function of a module does not trigger the  session attachable retrieving the secret automagically (which is expected by design): 
https://discord.com/channels/707636530424053791/1003718839739105300/1374464608416436326.

In this example, I indeed expect `exampleRef` to fail as I did not pass the token as a `GitOpts` -- which we special case when resolving git as a part of function parameters of a top-level module

**Comment by grouville** [comment]
I have a pending PR with the requested changes -- will push them after this week 🙏

**Comment by grouville** [comment]
@skycaptain I tried to push my branch with a few commits on top of your fork's main (as this is the reference of this PR) by force pushing (you only had a commit [that I have on my fork], so it felt safe) -- it worked to update the main branch's HEAD, but I'm getting errors atm, due to branch permissions. This updated your fork's main:
image

Can you please pick the commits from this branch: https://github.com/grouville/dagger/tree/with-authtoken and reopen the PR please ? I don't have the rights 🙏 Your commit is still present cc @jedevc

**Comment by skycaptain** [comment]
😮 Yeah, I see that force-pushing somehow removed my commit and might have auto-closed this PR. Since I don't have permission to reopen it, I created a new one: https://github.com/dagger/dagger/pull/10605. Since my commit was reverted anyway, starting with a fresh PR containing only our changes would have probably been easier though 😀.
